torch/csrc/api/include/torch/detail/TensorDataContainer.h

```diff
@@ -0,0 +1,293 @@
+//
+// At this point, theoretically there are two plausible ways for `{1}` to be matched to one of the
+// constructors of `TensorDataContainer`:
+//
+// 1. It can be a list-initialization of a scalar value, thus matching `TensorDataContainer(int value)`.
+// 2. It can be converted to `std::initializer_list<TensorDataContainer>`, thus matching
+//    `TensorDataContainer(std::initializer_list<TensorDataContainer>)`.
+//
+// How does the compiler decide which one to choose? According to
+// `https://en.cppreference.com/w/cpp/language/list_initialization`, braced-init-list always prefers
+// the constructor that takes `std::initializer_list`. Hence we happily move forward with constructor #2,
+// and it calls the following:
+//
+// `TensorDataContainer(1)`
+//
+// Now it matches `TensorDataContainer(int value)`, which stores `1` as a scalar value. All is good.
+struct TensorDataContainer {
+  // NOTE: For tensors with zero-size dimensions (e.g. `torch::tensor({{}, {}})`),
+  // the innermost empty braced-init-list `{}` matches the default constructor of
+  // the innermost `TensorDataContainer`.
+  TensorDataContainer() :
+      sizes_({0}),
+      scalar_type_(at::kFloat),
+      type_(TensorDataContainerType::InitList) {}
+#define TENSOR(T, S) \
+  TensorDataContainer(T value) : \
+      sizes_(), \
+      scalar_type_(at::k##S), \
+      type_(TensorDataContainerType::Scalar), \
+      scalar_(value) {}
+AT_FORALL_SCALAR_TYPES_AND3(Bool, Half, BFloat16, TENSOR)
+#undef TENSOR
+  TensorDataContainer(std::initializer_list<TensorDataContainer> init_list) :
+      sizes_(),
+      scalar_type_(init_list.begin()->scalar_type()),
+      type_(TensorDataContainerType::InitList),
+      init_list_(init_list) {
+    const TensorDataContainer& first_elem = *(init_list.begin());
+    for (const auto& elem : init_list) {
+      TORCH_CHECK(elem.sizes() == first_elem.sizes(),
+        "Expected all sub-lists to have sizes: ",
+        first_elem.sizes(),
+        " (e.g. ", first_elem, "), ",
+        "but got sub-list ",
+        elem,
+        " with sizes: ",
+        elem.sizes());
+      TORCH_CHECK(elem.scalar_type() == first_elem.scalar_type(),
+        "Expected all elements of the tensor to have the same scalar type: ",
+        first_elem.scalar_type(),
+        ", but got element of scalar type: ",
+        elem.scalar_type());
+    }
+    sizes_.reserve(first_elem.sizes().size() + 1);
+    sizes_.push_back(init_list.size());
+    sizes_.insert(sizes_.end(), first_elem.sizes().begin(), first_elem.sizes().end());
+  }
+
+#define TENSOR(T, S) \
+  TensorDataContainer(at::ArrayRef<T> values) : \
+      sizes_({(int64_t)values.size()}), \
+      scalar_type_(at::k##S), \
+      type_(TensorDataContainerType::Tensor) { \
+    at::AutoNonVariableTypeMode non_var_type_mode(true);  \
+    tensor_ = at::tensor(values, at::TensorOptions().device(at::kCPU).is_variable(false));       \
```

Comment:
What are the differences in semantics between at::tensor and torch::tensor? Does `at::tensor` only take in ArrayRefs?

Reply:
I believe `at::tensor` takes in scalar value / ArrayRef / std::vector / braced-init-list at the moment:
https://github.com/pytorch/pytorch/blob/d6f1e49c4a3d853edf18a53e827d648541fac44a/aten/src/ATen/templates/NativeFunctions.h#L30-L49

Reply:
ah, and at::tensor(1.) returns a 1D tensor instead of a scalar tensor. Do you think it would be worth cleaning up at::tensor to look more like torch::tensor at some point?

Reply:
Yes I just filed two issues on this topic:
https://github.com/pytorch/pytorch/issues/28929
https://github.com/pytorch/pytorch/issues/28930
It would be a good idea to unify the behavior of `at::tensor` and `torch::tensor`.